prusti-viper/src/encoder/mir/procedures/encoder/mod.rs

```diff
@@ -424,116 +427,277 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
-        self.encode_lft_assignments(block_builder, location, introduced_derived_lifetimes)?;
+        let new_derived_lifetimes = self.lifetimes.get_origin_contains_loan_at_mid(location);
+        block_builder.add_comment(format!("Prepare lifetimes for statement {:?}", location));
+        self.encode_lft(
+            block_builder,
+            location,
+            original_lifetimes,
+            derived_lifetimes,
+            &new_derived_lifetimes,
+        )?;
         Ok(())
     }
 
-    fn encode_end_lft(
+    /// Adds all statements needed for the encoding of the location.
+    fn encode_lft(
         &mut self,
         block_builder: &mut BasicBlockBuilder,
         location: mir::Location,
-        lifetimes: BTreeSet<String>,
+        old_original_lifetimes: &mut BTreeSet<String>,
+        old_derived_lifetimes: &mut BTreeMap<String, BTreeSet<String>>,
+        new_derived_lifetimes: &BTreeMap<String, BTreeSet<String>>,
     ) -> SpannedEncodingResult<()> {
-        for lifetime in lifetimes {
-            let lifetime_var = vir_high::VariableDecl::new(lifetime, vir_high::ty::Type::Lifetime);
+        let new_original_lifetimes: BTreeSet<String> = new_derived_lifetimes
+            .clone()
+            .into_values()
+            .flatten()
+            .collect();
+        self.encode_lft_return(
+            block_builder,
+            location,
+            old_derived_lifetimes,
+            new_derived_lifetimes,
+        )?;
+        self.encode_end_lft(
+            block_builder,
+            location,
+            old_original_lifetimes,
+            &new_original_lifetimes,
+        )?;
+        self.encode_dead_variables(
+            block_builder,
+            location,
+            old_derived_lifetimes,
+            new_derived_lifetimes,
+        )?;
+        self.encode_new_lft(
+            block_builder,
+            location,
+            old_original_lifetimes,
+            &new_original_lifetimes,
+        )?;
+        self.encode_lft_take(
+            block_builder,
+            location,
+            old_derived_lifetimes,
+            new_derived_lifetimes,
+        )?;
+
+        *old_original_lifetimes = new_original_lifetimes.clone();
+        *old_derived_lifetimes = new_derived_lifetimes.clone();
+        Ok(())
+    }
+
+    /// A variable can be killed if its lifetime was previously derived from lifetimes
+    /// but isn't anymore now.
+    fn variables_to_kill(
+        &mut self,
+        old_derived_lifetimes: &BTreeMap<String, BTreeSet<String>>,
+        new_derived_lifetimes: &BTreeMap<String, BTreeSet<String>>,
+    ) -> SpannedEncodingResult<BTreeSet<vir_high::VariableDecl>> {
+        let mut variables_to_kill = BTreeSet::new();
+        for derived_lifetime in old_derived_lifetimes.keys() {
+            if !new_derived_lifetimes.contains_key(derived_lifetime) {
+                if let Some((var, ty)) = self._procedure.get_var_of_lifetime(&derived_lifetime[..])
```

Comment:
```suggestion
                if let Some(local) = self.procedure.get_var_of_lifetime(&derived_lifetime[..])
```
I would suggest changing two things:
1. Since `procedure` is used now, remove the underscore from the name.
2. Change `get_var_of_lifetime` to return `mir::Local`.

Then, you should be able to replace the block with the following:
```rust
                     variables_to_kill.insert(self.encode_local(local));
```

Reply:
That is nicer, thanks! 
I also changed the type of the `Dead` statement from `VariableDecl` to `Expression`, is that ok?

Reply:
> I also changed the type of the `Dead` statement from `VariableDecl` to `Expression`, is that ok?

I think it should be fine.